File: code/operating_system/src/scheduling/first_come_first_serve/first_come_first_serve.java
import java.io.*; 
class FCFS 
{ 
 
public static void main(String args[]) throws Exception 
{ 
int n,at[],bt[],wt[],tat[]; 
float AWT=0; 
InputStreamReader isr=new InputStreamReader(System.in); 
BufferedReader br=new BufferedReader(isr); 
System.out.println(“Enter no of process”); 
n=Integer.parseInt(br.readLine()); 
  
wt=new int[n]; 
tat=new int[n]; 
bt=new int[n];
at=new int[n];
  
System.out.println(“Enter Burst time for each process\n”); 
for(int i=0;i<n;i++) 
{ 
System.out.println(“Process[“+(i+1)+"]"); 
bt[i]=Integer.parseInt(br.readLine()); 
} 
System.out.println(“\n\nEnter Around Time”); 
for(int i=0;i<n;i++) 
{ 
System.out.println(“Process[”+i+"]"); 
at[i]=Integer.parseInt(br.readLine()); 
} 
System.out.println(“\n”); 
wt[0]=0; 
for(int i=1;i<n;i++) 
{ 
wt[i]=wt[i-1]+bt[i-1]; 
wt[i]=wt[i]-at[i]; 
} 
for(int i=0;i<n;i++) 
{ 
tat[i]=wt[i]+bt[i]; 
awt=awt+wt[i]; 
} 
System.out.println(”  PROCESS\t\tBURST-TIME\tWAITING-TIME\tTURN AROUND-TIME\n“); 
for(int i=0;i<n;i++) 
{
System.out.println(”    “+ i + ”\t\t“+bt[i]+”\t“+wt[i]+”\t“+tat[i]);
} 
awt=awt/n; 
System.out.println(“\n”); 
System.out.println(“Avg waiting time=”+awt+”\n”); 
} 
} 

File: code/operating_system/src/scheduling/first_come_first_serve/first_come_first_serve.rs
fn find_waiting_time(bt: &[i32]) -> Vec<i32> {
    let mut wt = Vec::with_capacity(bt.len());
    // Waiting time for first process is 0
    wt.push(0);
    
    for i in 1..bt.len() {
        let last_wt = wt.last().cloned().unwrap();
        wt.push(bt[i - 1] + last_wt);
    }
    
    wt
}

fn find_turnaround_time(bt: &[i32], wt: &[i32]) -> Vec<i32> {
    let mut tat = Vec::with_capacity(bt.len());
    
    for i in 0..bt.len() {
        tat.push(bt[i] + wt[i]);
    }
    
    tat
}

fn find_avg_time(processes: &[i32], burst_times: &[i32]) {
    if processes.len() != burst_times.len() {
        return;
    }
    
    let n = processes.len();
    
    let wt = find_waiting_time(burst_times);
    
    let tat = find_turnaround_time(burst_times, &wt);
    
    println!("Processes\tBurst Time\tWaiting Time\t Turn around time");
    
    let total_wt = wt.iter().fold(0, |acc, &t| acc + t);
    let total_tat = tat.iter().fold(0, |acc, &t| acc + t);
    // Calculate total waiting ntime and total turn around time
    for i in 0..n {
        println!("{}\t{}\t{}\t{}", i + 1, burst_times[i], wt[i], tat[i]);
    }
    
    println!("Average waiting time: {}", total_wt / n as i32);
    println!("Average turn around time: {}", total_tat / n as i32);
}

fn main() {
    let processes = vec![1, 2, 3];
    let burst_time = vec![10, 5, 8];
    
    find_avg_time(&processes, &burst_time);
}
